pkg/controllers/agent/agent.go

```diff
@@ -17,31 +17,184 @@ package agent
 
-	clustermanagerv1alpha1 "github.com/font/onprem/api/v1alpha1"
+	v1alpha1 "github.com/font/onprem/api/v1alpha1"
 )
 
+var (
+	HeartBeatDelay = time.Second * 5
+	kubeAPIQPS     = 20.0
+	kubeAPIBurst   = 30
+
+	defaultConfigMapName      = "hub-cluster"
+	defaultSecretName         = "hub-cluster"
+	apiEndpointKey            = "server"
+	joinedClusterNameKey      = "joinClusterName"
+	joinedClusterNamespaceKey = "joinClusterNamespace"
+	caBundleKey               = "caBundle"
+	tokenKey                  = "token"
+)
+
+type JoinedClusterCoordinates struct {
+	APIEndpoint string
+	Name        string
+	Namespace   string
+	Token       []byte
+	CABundle    []byte
+}
+
+func GetJoinedClusterCoordinates(spokeClient client.Client) (*JoinedClusterCoordinates, error) {
+	namespaceBytes, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace")
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to get my namespace")
+	}
+	namespace := string(namespaceBytes)
+
+	// Extract contents from ConfigMap.
+	namespacedName := client.ObjectKey{Namespace: namespace, Name: defaultConfigMapName}
+	configMap := &corev1.ConfigMap{}
+	err = spokeClient.Get(context.TODO(), namespacedName, configMap)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to get configmap %+v", namespacedName)
+	}
+
+	apiEndpoint, found := configMap.Data[apiEndpointKey]
+	if !found || len(apiEndpoint) == 0 {
+		return nil, errors.Errorf("The configmap for the hub cluster is missing a non-empty value for %q", apiEndpointKey)
+	}
+
+	joinedClusterName, found := configMap.Data[joinedClusterNameKey]
+	if !found || len(joinedClusterName) == 0 {
+		return nil, errors.Errorf("The configmap for the hub cluster is missing a non-empty value for %q", joinedClusterNamespaceKey)
+	}
+
+	joinedClusterNamespace, found := configMap.Data[joinedClusterNamespaceKey]
+	if !found || len(joinedClusterNamespace) == 0 {
+		return nil, errors.Errorf("The configmap for the hub cluster is missing a non-empty value for %q", joinedClusterNamespaceKey)
+	}
+
+	// Extract contents from Secret.
+	namespacedName.Name = defaultSecretName
+	secret := &corev1.Secret{}
+	err = spokeClient.Get(context.TODO(), namespacedName, secret)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to get secret %+v", namespacedName)
+	}
+
+	token, found := secret.Data[tokenKey]
+	if !found || len(token) == 0 {
+		return nil, errors.Errorf("The secret for the hub cluster is missing a non-empty value for %q", tokenKey)
+	}
+
+	caBundle, found := secret.Data[caBundleKey]
+	if !found || len(caBundle) == 0 {
+		return nil, errors.Errorf("The secret for the hub cluster is missing a non-empty value for %q", caBundleKey)
+	}
+
+	return &JoinedClusterCoordinates{
+		APIEndpoint: apiEndpoint,
+		Name:        joinedClusterName,
+		Namespace:   joinedClusterNamespace,
+		Token:       token,
+		CABundle:    caBundle,
+	}, nil
+}
+
+func BuildHubClusterConfig(spokeClient client.Client) (*rest.Config, error) {
```

Comment:
This is minor, but you can pass the joinedClusterCoordinates into the BuildHubClusterConfig from the caller rather than calling it inside the function again.

Reply:
Yes, this was part of a refactoring so that I could always compare if the hub cluster coordinates changed in order to rebuild the config and client based on the updated cabundle/token/apiendpoint. However, it would end up requiring a bit of changes to the controller-runtime in order to plumb through a new config to the cache, mapper provider, etc. So it's not worth it for now and I will update this code accordingly.